Configurable signal handling, fix signal cleanup

michaeldwan opened the pull request:
*Issue #, if available:* #85 

*Description of changes:*
Adds a `DisableSignals` option to `machine.Config` for host apps that need to manage firecracker signals themselves. This also fixes an issue where signals weren't cleaned if the firecracker socket failed to open.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

xibz commented on machine.go:
```diff
@@ -880,3 +869,30 @@ func (m *Machine) waitForSocket(timeout time.Duration, exitchan chan error) erro
+
+// Set up a signal handler and pass INT, QUIT, and TERM through to firecracker
+func (m *Machine) setupSignals() {
+	if m.Cfg.DisableSignals {
```
How about instead of a `DisableSignals` flag, we have a strategy that allows for setting signals up. This strategy could be something like

```go
type BootstrapSignals func() error

type Machine struct {
    // machine field here
    BootstrapSignals BootstrapSignals
}
```

Then the default would be the `setupSignals` method you already have, but you can overwrite it with whatever you want.

michaeldwan replied:
It seems host apps need a way to control the list of signals to forward rather than inject new behavior. A bootstrap function is heavy handed for apps that just need to change which signals are forwarded since Machine doesn't export the command, so everything needs to be re-implemented. And for apps that don't want forwarding, an empty list seems clearer than a noop bootstrap function. 

I think exposing an optional list of signals to handle (which defaults to the current list) would make a cleaner API than injecting a function. What do you think?

xibz commented:
Also, please sign your commits with `git commit -s`. That seems to be the only reason why the CI is failing

michaeldwan commented:
@xibz I rebased with signed commits and updated the API to something explicit. How does this look?

Forward default signals:
```go
firecracker.NewMachine(ctx, firecracker.Config{})
```

Forward provided signals:
```go
firecracker.NewMachine(ctx, firecracker.Config{
    ForwardSignals: []os.Signal{os.Interrupt},
})
```

Disable all forwarding:
```go
firecracker.NewMachine(ctx, firecracker.Config{
    ForwardSignals: []os.Signal{},
})
```

michaeldwan commented:
FYI, we’ve been running this PR in production for a few hours through our custom nomad driver and disabling signals is working great for our use case.

xibz commented on machine.go:
```diff
@@ -880,3 +869,38 @@ func (m *Machine) waitForSocket(timeout time.Duration, exitchan chan error) erro
+func (m *Machine) setupSignals() {
+	signals := m.Cfg.ForwardSignals
+
+	if signals == nil {
```
Can we move this to `NewMachine`? That's where all defaults are occurring and I'd like to keep them in the same area.

michaeldwan replied:
Good call, fixed.

samuelkarp commented on machine.go:
```diff
@@ -880,3 +879,28 @@ func (m *Machine) waitForSocket(timeout time.Duration, exitchan chan error) erro
+	}
+
+	m.logger.Debugf("Setting up signal handler: %v", signals)
+	sigchan := make(chan os.Signal)
```
Can you make `sigchan` have length greater than or equal to `len(signals)`?  From [the `signal.Notify` documentation](https://golang.org/pkg/os/signal/#Notify):

> Package signal will not block sending to c: the caller must ensure that c has sufficient buffer space to keep up with the expected signal rate. For a channel used for notification of just one signal value, a buffer of size 1 is sufficient.

michaeldwan commented:
The signal channel is now buffered. I'm going to try and get a test for this. I can't get the tests to run correctly on my mac so I'll be debugging on BuildKite... ignore failure spam for a bit.

michaeldwan commented:
I've made the requested changes and have a test ensuring signals are forwarded correctly. Root tests are failing before they start, but I don't think it's related to this change. Let me know how it looks.

samuelkarp commented:
@michaeldwan Thank you for being so persistent here!  You are correct that the root test issue is not related to your change.  I opened https://github.com/firecracker-microvm/firecracker-go-sdk/issues/168 to track the root cause and https://github.com/firecracker-microvm/firecracker-go-sdk/pull/169 should address it for now.  Can you rebase on top of `master` to pick up that change?  That should fix the issue you had and allow the root tests to run properly for your PR.

samuelkarp commented:
The test looks good to me too.  Once we have this rebased (or, if you don't have the time, I can do this for you) we'll be good to go.

michaeldwan commented:
@samuelkarp I rebased and tests are passing, thanks for the fix

xibz commented:
@michaeldwan - Thank you for the contribution and persistence. These changes look great. Going to go ahead and merge.
